nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,433 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+ 
+  mkExcludeFile = cfg:
+    # Write each exclude pattern to a new line
+    pkgs.writeText "excludefile" (concatStringsSep "\n" cfg.exclude);
+
+  mkPathsArgs = cfg:
+    # Shell-escape each path and put spaces between them
+    concatMapStringsSep " " escapeShellArg cfg.paths;
+
+  mkArchiveArg = cfg:
+    # The repo URL can be left out, 
+    # because it is given in $BORG_REPO
+    escapeShellArg "${cfg.repo}::${cfg.archiveName}";
+
+  mkKeepArgs = cfg:
+    # If cfg.prune.keep e.g. has a yearly attribute,
+    # its content is passed on as --keep-yearly
+    concatStringsSep " "
+      (mapAttrsToList (x: y: "--keep-${x}=${toString y}") cfg.prune.keep);
+
+  mkBackupScript = cfg: ''
+    ${cfg.preHook}
+    borg init \
+      --encryption ${cfg.encryption.mode} \
+      || true
```

Comment:
Is it difficult to run init only when needed, and fail on error? This seems to ignore errors, which gives me a bad feeling.

Reply:
It is better to touch a file after `borg init` and only call `borg init` if this file does not exist.

Reply:
I think it is better to call e.g. `borg list` to check if a repo exists. This will also cover the case where the repo has been created manually.

Reply:
I think borg has some safety measures in case somebody creates backup on an external drive/remote filesystem and the mount is currently not active to avoid making backups on the underlying device.
Can we cover this as well?

Reply:
The reason is that user might thing their backup works, while they backup the whole time on the same filesystem where the data is located.

Reply:
I don't know of such measures. Maybe adding a `doInit` option is a good idea?